README.md
**Version**: 0.9.1  
**Author**: Hannah Weller  
**Email**: hannahiweller@gmail.com  
**GitHub**: https://github.com/hiweller  

An R package that counts colors within color range(s) in images, and
    provides a masked version of the image with targeted pixels
    changed to a different color. Output includes the locations
    of the pixels in the images, and the proportion of the image
    within the target color range with optional background masking.
    Users can specify multiple color ranges for masking.

For more information, read the [introduction](https://cran.r-project.org/package=countcolors/vignettes/Introduction.html) on CRAN.

## About countcolors

This package is the result of a collaboration between  [Sarah
Hooper](https://scholar.google.com/citations?user=gaUr5yEAAAAJ&hl=en),
[Sybill Amelon](https://www.nrs.fs.fed.us/people/Amelon), and me
([Hannah
Weller](https://scholar.google.com/citations?user=rjI5wpEAAAAJ&hl=en)),
with the goal of quantifying the area of white-nose syndrome infection
of bat wings.

Countcolors allows users to quantify regions of an image by color(s).
It's a companion package to
[colordistance](https://CRAN.R-project.org/package=colordistance), which
provides quantitative color comparisons of images. While both of these
packages borrow techniques from image processing, they don't use any
machine learning, adaptive thresholding, or object detection. This makes
them reliable and easy to use, but limited in application.

Countcolors is best suited for images where the color(s) of interest are
reasonably different from the rest of the image – so if you've got an
image of some treetops, it can quantify the amount of green tree cover,
but it wouldn't necessarily be able to distinguish a species with leaves
of one slightly different shade of green from all the others due to
variations in lighting (among other sources of variation). If this is
the kind of specificity you need, you're better off just mucking in
and using real image processing. I recommend the [OpenCV
library](https://opencv.org/) for C++ or Python.

**December 27, 2018**: Fixed bugs in `countColors` that threw errors when
specifying filepaths for `save.indicator` or providing numeric values for
`target.color`. Thanks to Lee Yuhui for discovering them.

## Installation instructions

1. Install the `devtools` package (`install.packages("devtools")`).
2. Install `countcolors` using the `install_github` function:
```{r}
devtools::install_github("hiweller/countcolors")
```

## Questions or feedback?

Email me: <hannahiweller@gmail.com>


NEWS.md
# countcolors 0.9.1
Fixed some bugs with countColors accepting RGB matrices for indicator color specification. Fixed redundant vignette names.

# countcolors 0.9.0

First release version.


R/search_color_ranges.R
# Find pixels in an image that fall inside of a color range

# rectangularRange: find pixels within a target range defined by boundaries in
# each channel

#' Find pixels within a target color range defined by boundaries in each channel
#'
#' Searches for pixels within a set of upper and lower limits for each color
#' channel. Essentially draws a 'box' around a region of color space in which to
#' search for pixels.
#'
#' @param pixel.array An image represented as a 3D array (as read in by
#'   \code{\link[jpeg]{readJPEG}}, \code{\link[png]{readPNG}}, or
#'   \code{\link[colordistance]{loadImage}}) in which to change pixel colors.
#'
#' @param upper,lower RGB triplet specifying the bounds of color space search.
#'   See details.
#'
#' @param target.color Color with which to replace specified pixels. Can be
#'   either an RGB triplet or one of the colors listed by
#'   \code{\link[grDevices]{colors}}.
#'
#' @param main Optional title to display for image.
#'
#' @param plotting Logical. Should output be plotted in the plot window?
#'
#' @param color.pixels Logical. Should a diagnostic image with pixels changed to
#'   \code{target.color} be returned?
#'
#' @return A list with the following elements:
#' \itemize{
#'   \item \code{pixel.idx}: Coordinates of pixels within color range.
#'   \item \code{img.fraction}: Proportion of the image within color range.
#'   \item \code{original.img}: The original RGB array.
#'   \item \code{indicator.img}: If \code{color.pixels = TRUE}, RGB array with
#'   color-swapped pixels.
#' }
#'
#' @examples
#' flowers <- jpeg::readJPEG(system.file("extdata", "flowers.jpg", package =
#' "countcolors"))
#'
#' # Define upper and lower bounds for white
#' lower <- rep(0.8, 3)
#' upper <- rep(1, 3)
#'
#' white.flowers <- countcolors::rectangularRange(flowers, rep(1, 3), rep(0.85,
#' 3), target.color = "turquoise")
#'
#' @details
#' \code{lower} and \code{upper} should be vectors of length 3 in a 0-1 range,
#' in the order R-G-B. For example, the upper bounds for white would be c(1, 1,
#' 1), and the lower bounds might be c(0.8, 0.8, 0.8). This would search for all
#' pixels where the red value, blue value, AND green value are all between 0.8
#' and 1.
#'
#' @export
rectangularRange <- function(pixel.array, upper, lower,
                             target.color = "green", main = "",
                             color.pixels = TRUE, plotting = TRUE) {

  # Find all pixels within target range + get their locations
  idx <- which( (lower[1] <= pixel.array[, , 1] &
                  pixel.array[, , 1] <= upper[1]) &
                 (lower[2] <= pixel.array[, , 2] &
                    pixel.array[, , 2] <= upper[2]) &
                 (lower[3] <= pixel.array[, , 3] &
                    pixel.array[, , 3] <= upper[3]),
               arr.ind = TRUE)

  # Create list to return, including index of pixels within range, number of
  # pixels in range, fraction of image in range, and original image
  return.list <- list(pixel.idx = idx, pixel.count = nrow(idx),
                      img.fraction =
                        nrow(idx) / (nrow(pixel.array) * ncol(pixel.array)),
                      original.img = pixel.array)

  # If color.pixels is TRUE, create a version of the image with the pixels
  # changed to a diagnostic color and add indicator image onto returned list
  if (color.pixels | plotting) {
    indicator.img <-
      countcolors::changePixelColor(pixel.array = pixel.array,
                                      pixel.idx = idx,
                                      target.color = target.color,
                                      plotting = FALSE, return.img = TRUE)
    return.list$indicator.img <- indicator.img

    # If plotting = TRUE, plot indicator.img in the plot window
    if (plotting) {
      countcolors::plotArrayAsImage(indicator.img, main = main)
    }

  }

  return(return.list)

}



#' Find pixels within a target color range defined by a center and search radius
#'
#' Searches for pixels within a radius of a single color. Draws a sphere around
#' a region of color space in which to search for pixels.
#'
#' @param pixel.array An image represented as a 3D array (as read in by
#'   \code{\link[jpeg]{readJPEG}}, \code{\link[png]{readPNG}}, or
#'   \code{\link[colordistance]{loadImage}}) in which to change pixel colors.
#'
#' @param center A single color (RGB triplet) around which to search. RGB range
#'   0-1 (not 0-255).
#'
#' @param radius A value between 0 and 1 specifying the size of the area around
#'   \code{center} to search.
#'
#' @param target.color Color with which to replace specified pixels. Can be
#'   either a an RGB triplet or one of the colors listed by
#'   \code{\link[grDevices]{colors}}.
#'
#' @param main Optional title to display for image.
#'
#' @param plotting Logical. Should output be plotted in the plot window?
#'
#' @param color.pixels Logical. Should a diagnostic image with pixels changed to
#'   \code{target.color} be returned?
#'
#' @return A list with the following elements:
#' \itemize{
#'   \item \code{pixel.idx}: Coordinates of pixels within color range.
#'   \item \code{img.fraction}: Proportion of the image within color range.
#'   \item \code{original.img}: The original RGB array.
#'   \item \code{indicator.img}: If \code{color.pixels = TRUE}, RGB array with
#'   color-swapped pixels.
#' }
#'
#' @examples
#' # Target color: change all of the red flowers to turquoise
#' flowers <- jpeg::readJPEG(system.file("extdata", "flowers.jpg", package =
#' "countcolors"))
#'
#' # Red:
#' center <- c(255, 75, 75) / 255
#'
#' # Setting the radius too low:
#' red.flowers <- countcolors::sphericalRange(flowers, center = center, radius =
#' 0.05, target.color = "turquoise")
#'
#' # Setting the radius too high:
#' red.flowers <- countcolors::sphericalRange(flowers, center = center, radius =
#' 0.4, target.color = "turquoise")
#'
#' # Setting the radius just right:
#' red.flowers <- countcolors::sphericalRange(flowers, center = center, radius =
#' 0.2, target.color = "turquoise")
#'
#' @details
#' \code{lower} and \code{upper} should be vectors of length 3 in a 0-1 range,
#' in the order R-G-B. For example, the upper bounds for white would be c(1, 1,
#' 1), and the lower bounds might be c(0.8, 0.8, 0.8). This would search for all
#' pixels where the red value, blue value, AND green value are all between 0.8
#' and 1.
#'
#' @export
sphericalRange <- function(pixel.array, center, radius,
                           target.color = "green", main = "",
                           color.pixels = TRUE, plotting = TRUE) {

  # Assuming RGB, change radius percent to fraction of maximum distance in RGB
  # space
  radius <- radius * sqrt(sum( (rep(0, 3) - rep(1, 3)) ^ 2))

  pixel.distances <- matrix(NA, nrow = dim(pixel.array)[1],
                            ncol = dim(pixel.array)[2])

  # For every pixel, calculate distance from center
  for (i in 1:dim(pixel.array)[1]) {
    for (j in 1:dim(pixel.array)[2]) {
      pixel.distances[i, j] <- sqrt(sum( (pixel.array[i, j, ] - center) ^ 2))
    }
  }

  # Index every pixel with distance <= radius
  idx <- which(pixel.distances <= radius, arr.ind = TRUE)

  # Provide warning if no pixels were indexed at all
  if (nrow(idx) == 0) {
    warning("No pixels within color range in provided image")
  }

  # Create list to return, including index of pixels within range, number of
  # pixels in range, fraction of image in range, and original image
  return.list <- list(pixel.idx = idx,
                      pixel.count = nrow(idx),
                      img.fraction = nrow(idx) /
                        (nrow(pixel.array) * ncol(pixel.array)),
                      original.img = pixel.array)

  # If color.pixels is TRUE, create a version of the image with the pixels
  # changed to a diagnostic color and add indicator image onto returned list
  if (color.pixels | plotting) {
    indicator.img <-
      countcolors::changePixelColor(pixel.array = pixel.array,
                                      pixel.idx = idx,
                                      target.color = target.color,
                                      plotting = FALSE, return.img = TRUE)
    return.list$indicator.img <- indicator.img

    # If plotting = TRUE, plot indicator.img in the plot window
    if (plotting) {
      countcolors::plotArrayAsImage(indicator.img)
    }

  }

  return(return.list)

}


R/count_colors_in_images.R
#' Count the number of pixels within a color range or ranges
#'
#' Counts the pixels within a color range; ranges can be spherical (see
#' \code{\link{sphericalRange}}) or rectangular
#' (\code{\link{rectangularRange}}). If multiple ranges are specified, each one
#' is colored using a different indicator color.
#'
#' @param path Path to the image (JPEG or PNG).
#'
#' @param color.range Type of range being specified. Must be either "spherical"
#'   or "rectangular".
#'
#' @param center A vector or n x 3 matrix of color centers (RGB triplets) around
#'   which to search using spherical color range. RGB range 0-1 (not 0-255). See
#'   details.
#'
#' @param radius Values between 0 and 1 specifying the size of the area around
#'   \code{center} to search. The same number of centers and radii must be
#'   specified.
#'
#' @param lower,upper  RGB triplet(s) specifying the bounds of color space
#'   to search. Must be the same length. See details.
#'
#' @param bg.lower,bg.upper RGB triplets specifying the bounds of color space to
#'   ignore as background, or \code{NULL} to use the entire image.
#'
#' @param target.color If an indicator image is created, the color with which to
#'   replace specified pixels. Can be either an RGB triplet or one of the
#'   colors listed by \code{\link[grDevices]{colors}}.
#'
#' @param plotting Logical. Should output be plotted in the plot window?
#'
#' @param save.indicator Logical. If TRUE,
#'   saves image to the same directory as the original image as
#'   'originalimagename_masked.png'; if a path is provided, saves it to that
#'   directory/name instead, also as a PNG.
#'
#' @param dpi Resolution (dots per image) for saving indicator image.
#'
#' @param return.indicator Logical. Should an indicator image (RGB array with
#'   targeted pixels changed to indicator color) be returned?
#'
#' @return A list with the following attributes:
#' \itemize{
#'   \item \code{pixel.idx}: Unique coordinates of pixels within color range(s).
#'   \item \code{pixel.fraction}: Proportion of the non-background image within
#'   color range(s), found by dividing the number of pixels in \code{pixel.idx}
#'   by the number of non-background pixels in the image.
#'   \item \code{indicator.img}: If \code{return.indicator = TRUE}, RGB array with
#'   color-swapped pixels.
#' }
#'
#' @seealso \code{\link{countColorsInDirectory}}
#'
#' @examples
#'
#' # Try out a few different radii for white pelicans
#' pelicans <- system.file("extdata", "pelicans.jpg", package = "countcolors")
#'
#' white.ctr <- rep(0.9, 9)
#' white.radii <- c(0.5, 0.3, 0.1)
#'
#' # Magenta = 50% threshold, cyan = 30% threshold, yellow = 10% threshold
#' pelican.example <- countcolors::countColors(pelicans, center = white.ctr,
#' radius = white.radii, bg.lower = NULL, plotting = TRUE)
#'
#' @details More than one set of ranges can be specified for the color search
#'   space, but the same number of arguments must be provided in each case (so
#'   the number of pixels and centers must be the same if using a spherical
#'   range, and the number of upper and lower bounds must be the same if using a
#'   rectangular one).
#'
#' For \code{center}, \code{upper}, and \code{lower}, which call for RGB
#' triplets, provide either a vector of RGB triplets in RGB order, i.e. c(R1,
#' G1, B1, R2, G2, B2, etc) or a 3-column matrix with one row per RGB triplet.
#' If a vector is provided, it is coerced to a 3-column matrix, and must
#' therefore be a multiple of 3.
#'
#'
#' @export
countColors <- function(path, color.range = "spherical",
                       center, radius, lower, upper,
                       bg.lower = rep(0.8, 3), bg.upper = rep(1, 3),
                       target.color = c("magenta", "cyan", "yellow"),
                       plotting = FALSE, save.indicator = FALSE, dpi=72,
                       return.indicator = FALSE) {

  # Takes a path to an image and loads it
  img <- colordistance::loadImage(path, lower = bg.lower, upper = bg.upper)
  original <- img$original.rgb

  # If saving is on but no path was specified, set it to same directory as
  # original image
  if (isTRUE(save.indicator)) {
    destination <- paste(tools::file_path_sans_ext(path),
                         "_masked.png", sep = "")
  } else if (is.character(save.indicator)) {
    # If save.indicator is a filepath, save as a png and set save.indicator to a
    # logical value
    if (dir.exists(save.indicator)) {
      destination <- paste(save.indicator, tools::file_path_sans_ext(path),
                           ".png", sep = "")
    } else {
      destination <- paste(tools::file_path_sans_ext(save.indicator),
                           ".png", sep = "")
    }
    save.indicator <- TRUE
  }

  # If any output that requires colored indicator image is flagged, set
  # get.indicator to TRUE
  if (plotting | return.indicator | save.indicator) {
    get.indicator <- TRUE

    # Convert target.color to hex codes for indicator image
    # If target.color is color names, convert to RGB matrix
    if (is.character(target.color)) {
      # convert color names to RGB triplets
      target.color <- t(grDevices::col2rgb(target.color)) / 255
      # If vector or matrix, convert to/check for 3-column matrix
    } else if (is.numeric(target.color)) {
      # If a vector, convert to a matrix
      if (is.vector(target.color)) {
        if (length(target.color) %% 3 == 0) {
          target.color <- matrix(target.color, ncol = 3, byrow = TRUE)
        } else {
          stop(paste("Target.color must be a vector of color names",
                     "a vector of RGB triplets (multiple of 3)",
                     "or a matrix with target.colors as rows",
                     sep = ", "))
          # If already a matrix, check that it has 3 columns
        }
      } else if (is.matrix(target.color)) {
        if (ncol(target.color) != 3) {
          stop("target.colors matrix must have 3 columns, 1 per channel")
        }
      }
    }

    # Convert to hex colors
    target.color <- apply(target.color, 1,
                    function(i) grDevices::rgb(i[1], i[2], i[3]))
  } else {
    get.indicator <- FALSE
  }

  # If range = "spherical", check that center and radius are specified and are
  # coercible to appropriate formats
  if (tolower(color.range) == "spherical") {

    # Check that center and radius were specified and are of appropriate types
    if (!exists("center") | !exists("radius")) {
      stop("Center and radius must be specified for a spherical color range")
    }

    # Make sure that center is of appropriate format (divisible by 3/matrix with
    # 3 columns)
    if (is.vector(center)) {
      if (length(center) %% 3 == 0) {
        center <- matrix(center, ncol = 3, byrow = TRUE)
      } else {
        stop("Center must be a vector with length a multiple of 3 or a matrix
             of centers")
      }
    } else if (is.matrix(center)) {
      if (ncol(center) != 3) {
        stop("Centers matrix must have 3 columns, 1 per channel")
      }
    }

    # And make sure everything is in the 0-1 range; if not, assume 0-255 and
    # divide by 255
    if (range(center)[2] > 1) {
      center <- center / 255
    }

    # Make sure that an appropriate number of radii were provided
    if (nrow(center) != length(radius)) {
      stop("Number of centers and radii differ")
    }

    # If everything checks out, print the centers/radii being screened
    message("Using spherical range(s):")
    for (i in 1:nrow(center)) {
      message(paste("Center: ", paste(center[i, ], collapse = ", "),
                    " +/- ", radius[i] * 100, "%", sep = ""))
    }

    # Make sure there are enough colors to use for color indexing; if not, just
    # repeat color vector
    if (length(target.color) < nrow(center)) {
      target.color <- rep(target.color, round(nrow(center) /
                                                length(target.color)) + 1)
    }

    # Get range using first color
    filtered.img <- sphericalRange(pixel.array = original, center = center[1, ],
                                   radius = radius[1], plotting = FALSE,
                                   color.pixels = get.indicator,
                                   target.color = target.color[1])
    idx <- filtered.img$pixel.idx

    # Keep the indicator image (color inside pixel range changed to indicator
    # color) and pixel indices for later display
    if (get.indicator) {
      indicator.img <- filtered.img$indicator.img
    }


    # If more than one center/radius pair were specified, screen for each color
    # in turn, but do the color indexing on the same image so we can see
    # everything that is being counted
    if (nrow(center) > 1) {
      for (i in 2:nrow(center)) {
        filtered.img <- sphericalRange(pixel.array = original,
                                       center = center[i, ], radius = radius[i],
                                       plotting = FALSE, color.pixels = FALSE)
        if (get.indicator) {
          indicator.img <- changePixelColor(indicator.img,
                                            filtered.img$pixel.idx,
                                            target.color = target.color[i],
                                            return.img = TRUE, plotting = FALSE)
        }
        idx <- rbind(idx, filtered.img$pixel.idx)
      }
    }

  } else if (tolower(color.range) == "rectangular") {
    # If range is set to rectangular, perform similar checks

    # Check that lower and upper ranges were specified
    if (!exists("lower") | !exists("upper")) {
      stop("Lower and upper ranges must be specified for
           a rectangular color range")
    }

    # Make sure that lower bounds are of appropriate format (divisible by
    # 3/matrix with 3 columns)
    if (is.vector(lower)) {
      if (length(lower) %% 3 == 0) {
        lower <- matrix(lower, ncol = 3, byrow = TRUE)
      } else {
        stop("Lower must be a vector with length a multiple of 3
             or a matrix of lower bounds")
      }
    } else if (is.matrix(lower)) {
      if (ncol(lower) != 3) {
        stop("Lower bounds matrix must have 3 columns, 1 per channel")
      }
    }

    # Same with upper; there is probably a more elegant way to do this than just
    # doing both separately... probably with quote/eval, but the variable
    # reassignment makes me itchy
    if (is.vector(upper)) {
      if (length(upper) %% 3 == 0) {
        upper <- matrix(upper, ncol = 3, byrow = TRUE)
      } else {
        stop("Upper must be a vector with length a multiple of 3
             or a matrix of lower bounds")
      }
    } else if (is.matrix(upper)) {
      if (ncol(upper) != 3) {
        stop("Upper bounds matrix must have 3 columns, 1 per channel")
      }
    }

    # Make sure that the same number of upper and lower bounds were provided
    if (nrow(lower) != nrow(upper)) {
      stop("Number of lower and upper bounds differs")
    }

    # And make sure everything is in the 0-1 range; if not, assume 0-255 and
    # divide by 255
    if (range(lower)[2] > 1 | range(upper)[2] > 1) {
      lower <- lower / 255
      upper <- upper / 255
    }

    message("Using rectangular bound(s):")
    for (i in 1:nrow(lower)) {
      message(paste("R: ", paste(lower[i, 1], "-", upper[i, 1], sep = ""),
                    "; G: ", paste(lower[i, 2], "-", upper[i, 2], sep = ""),
                    "; B: ", paste(lower[i, 3], "-", upper[i, 3], sep = ""),
                    sep = ""))
    }

    # Get range using first color
    filtered.img <- rectangularRange(pixel.array = original, lower = lower[1, ],
                                     upper = upper[1, ], plotting = FALSE,
                                     color.pixels = get.indicator,
                                     target.color = target.color[1])
    idx <- filtered.img$pixel.idx

    if (get.indicator) {
      indicator.img <- filtered.img$indicator.img
    }

    # If more than one lower/upper bound pair were specified, repeat the process
    # over increasingly filtered images
    if (nrow(lower) > 1) {
      for (i in 2:nrow(lower)) {
        filtered.img <- rectangularRange(pixel.array = original,
                                         lower = lower[i, ], upper = upper[i, ],
                                         plotting = FALSE, color.pixels = FALSE)
        if (get.indicator) {
          indicator.img <- changePixelColor(indicator.img,
                                            filtered.img$pixel.idx,
                                            target.color = target.color[i],
                                            return.img = TRUE, plotting = FALSE)
        }
        idx <- rbind(idx, filtered.img$pixel.idx)
      }
    }



  }
  else {
    stop("color.range must be either 'spherical' or 'rectangular'")
  }

  # Plot if indicated
  if (plotting) {
    countcolors::plotArrayAsImage(indicator.img)
  }

  # If destination specified (save.indicator was either TRUE or a path), save a
  # comparison image with original and masked images
  if (exists("destination")) {
    png::writePNG(indicator.img, target = destination, dpi = dpi)
    message(paste("Output image written to", destination))
  }

  return.list <- list(pixel.idx = unique(idx),
                      pixel.fraction =
                        nrow(unique(idx)) / nrow(img$filtered.rgb.2d))

  if (return.indicator) {
    return.list$indicator.img <- indicator.img
  }

  return(return.list)

}

# countColorInDirectory: wrapper for countColor, applies to every picture in a
# folder and returns a dataframe with % cover by image
#
# folder: folder with images (JPG/PNG) in it
# ... : arguments passed to 'countColor' for each image
#'
#' Count colors within range(s) in every image in a directory
#'
#' A wrapper for \code{\link{countColors}} that finds every image (JPEG or PNG) in a #' folder and counts colors in each image.
#'
#' @param folder Path to a folder containing images.
#'
#' @inheritParams countColors
#'
#' @return A list of \code{\link{countColors}} lists, one for each image.
#'
#' @seealso \code{\link{countColors}}
#'
#' @examples
#' \dontrun{
#' folder <- system.file("extdata", package = "countcolors")
#'
#' # Screen out white in both the flower image and the pelican image
#' upper <- c(1, 1, 1)
#' lower <- c(0.8, 0.8, 0.8)
#'
#' white.screen <- countcolors::countColorsInDirectory(folder, color.range =
#' "rectangular", upper = upper, lower = lower, bg.lower = NULL, plotting =
#' TRUE, target.color = "turquoise")
#'}
#' @export
countColorsInDirectory <- function(folder,
                                   color.range = "spherical",
                                   center, radius, lower, upper,
                                   bg.lower = rep(0.8, 3), bg.upper = rep(1, 3),
                                   target.color = c("magenta", "cyan", "yellow"),
                                   plotting = FALSE, save.indicator = FALSE, dpi=72,
                                   return.indicator = FALSE) {

  # Get a list of all the images
  images <- colordistance::getImagePaths(folder)

  # For each image, run countColors function and store the result
  output <- vector("list", length = length(images))

  # Run the first one without suppressing messages to screen for color ranges
  message(length(images), " images \n")
  output[[1]] <-
    countcolors::countColors(images[1], color.range = color.range,
      center = center, radius = radius,
      lower = lower, upper = upper,
      bg.lower = bg.lower, bg.upper = bg.upper,
      target.color = target.color, plotting = plotting,
      save.indicator = save.indicator, dpi = dpi,
      return.indicator = return.indicator)

  message("\n Image: ", basename(images[1]))

  for (i in 2:length(images)) {
    message("Image: ", basename(images[i]))
    output[[i]] <-suppressMessages(countcolors::countColors(images[i],
                      color.range = color.range,
                      center = center, radius = radius,
                      lower = lower, upper = upper,
                      bg.lower = bg.lower, bg.upper = bg.upper,
                      target.color = target.color, plotting = plotting,
                      save.indicator = save.indicator, dpi = dpi,
                      return.indicator = return.indicator))
  }

  names(output) <- tools::file_path_sans_ext(basename(images))

  return(output)

}


R/visualization_tools.R
# Functions for visualizing pixel count output

#' Plot a 3D array as an RGB image
#'
#' Plots a 3D array as an RGB image in the plot window.
#'
#' @param rgb.array 3D RGB array with R, G, and B channels (pixel rows x pixel
#'   columns x color channels) to be plotted in the plot window as an actual
#'   image.
#'
#' @param main Optional title to display for image.
#'
#' @examples
#' # Read in image
#' flowers <- jpeg::readJPEG(system.file("extdata", "flowers.jpg", package =
#' "countcolors"))
#'
#' # Plot
#' plotArrayAsImage(flowers, main = "flowers!")
#'
#' @export
plotArrayAsImage <- function(rgb.array, main = "") {

    # Make sure the array is 3-dimensional

    # If there are more than 3 channels/slices in the 3rd dimension (i.e. an
    # alpha channel as 4th), only use the first 3
    if (length(dim(rgb.array)) != 3) {
        stop("RGB_array must be an array of three dimensions (pixel rows,
             pixel columns, and color channels)")
    } else if (dim(rgb.array)[3] != 3) {
      warning("Provided array has more than 3 channels; using only the
                first 3 as R, G, and B channels")
    }

    # Change graphical parameters for image display
    op <- graphics::par(mar = c(0, 0, 2, 0))
    asp <- dim(rgb.array)[1] / dim(rgb.array)[2]

    # Initialize empty plot window
    graphics::plot(0:1, 0:1, type = "n", ann = F, axes = F, asp = asp)

    # Use rasterImage to actually plot the image
    graphics::rasterImage(rgb.array, 0, 0, 1, 1)
    graphics::title(main, line = 0)

    # Return to original graph window settings
    graphics::par(op)
}



#' Change all specified pixels to a new color
#'
#' Changes pixels in an image to a different color and displays and/or returns
#' the re-colored image.
#'
#' @param pixel.array An image represented as a 3D array (as read in by
#'   \code{\link[jpeg]{readJPEG}}, \code{\link[png]{readPNG}}, or
#'   \code{\link[colordistance]{loadImage}}) in which to change pixel colors.
#'
#' @param pixel.idx An n x 2 matrix of index coordinates specifying which pixels
#'   to change, where rows are pixels and columns are X and Y coordinates in the
#'   image.
#'
#' @param target.color Color with which to replace specified pixels. Can be
#'   either a an RGB triplet or one of the colors listed by
#'   \code{\link[grDevices]{colors}}.
#'
#' @param return.img Logical. Should RGB array with swapped colors be returned?
#'
#' @param plotting Logical. Should output be plotted in the plot window?
#'
#' @param main Optional title to display for image.
#'
#' @return Raster array with indicated pixels changed to target color, if
#'   \code{return.img = TRUE}.
#'
#' @examples
#' # Change a rectangle in the center to black
#' flowers <- jpeg::readJPEG(system.file("extdata", "flowers.jpg", package =
#' "countcolors"))
#'
#' sinister.object <- expand.grid(c(114:314), c(170:470))
#'
#' countcolors::changePixelColor(flowers, sinister.object, target.color = "black")
#'
#'\dontrun{
#' # Change all the white flowers to magenta
#' indicator.img <- countcolors::sphericalRange(flowers, center = c(1, 1, 1),
#' radius = 0.1, color.pixels = TRUE, plotting = FALSE)
#'
#' countcolors::changePixelColor(flowers, indicator.img$pixel.idx,
#' target.color="magenta")
#' }
#' @export
changePixelColor <- function(pixel.array, pixel.idx, target.color = "green",
                             return.img = FALSE, plotting = TRUE, main = "") {

    # Make sure target color is or can be coerced to an RGB triplet
    if (is.character(target.color)) {
        target.color <- as.vector(grDevices::col2rgb(target.color) / 255)
    }

    # Check that vector is of appropriate length and has a 0-1 (not 0-255) range
    if (length(target.color) != 3) {
        stop("'target.color' must be a numeric vector of length 3 with
             values between 0 and 1 or one of the colors listed by colors()")
    } else if (range(target.color)[2] > 1) {
        target.color <- target.color / 255
    }

    # Change specified pixels to target color
    for (i in 1:nrow(pixel.idx)) {
        pixel.array[pixel.idx[i, 1], pixel.idx[i, 2], ] <- target.color
    }

    # Plot if plotting = TRUE
    if (plotting) {
        countcolors::plotArrayAsImage(pixel.array, main = main)
    }

    # Return image with changed pixel colors if return.img = TRUE
    if (return.img) {
        return(pixel.array)
    }

}
